File: app/chunking/document_chunker.py
"""Chunk documents with overlap for better retrieval"""
import logging
from typing import List

from app.schemas.documents import Document, Chunk

logger = logging.getLogger(__name__)


class DocumentChunker:
    """Chunk documents with overlap for better retrieval"""
    
    def __init__(
        self,
        chunk_size: int = 800,
        chunk_overlap: int = 200,
        separator: str = "\n\n"
    ):
        self.chunk_size = chunk_size
        self.chunk_overlap = chunk_overlap
        self.separator = separator
    
    async def chunk_document(self, document: Document) -> List[Chunk]:
        """Chunk a document into smaller pieces"""
        text = document.content
        
        # Split by separator first
        splits = self._split_text(text, self.separator)
        
        chunks = []
        current_chunk = []
        current_length = 0
        
        for i, split in enumerate(splits):
            split_length = len(split)
            
            if current_length + split_length > self.chunk_size and current_chunk:
                # Create chunk
                chunk_text = self.separator.join(current_chunk)
                chunk = self._create_chunk(
                    chunk_text,
                    document,
                    chunk_index=len(chunks)
                )
                chunks.append(chunk)
                
                # Start new chunk with overlap
                overlap_length = 0
                overlap_parts = []
                for part in reversed(current_chunk):
                    if overlap_length + len(part) < self.chunk_overlap:
                        overlap_parts.insert(0, part)
                        overlap_length += len(part)
                    else:
                        break
                
                current_chunk = overlap_parts
                current_length = overlap_length
            
            current_chunk.append(split)
            current_length += split_length
        
        # Add final chunk
        if current_chunk:
            chunk_text = self.separator.join(current_chunk)
            chunk = self._create_chunk(
                chunk_text,
                document,
                chunk_index=len(chunks)
            )
            chunks.append(chunk)
        
        return chunks
    
    def _split_text(self, text: str, separator: str) -> List[str]:
        """Split text by separator"""
        if separator:
            splits = text.split(separator)
        else:
            splits = list(text)
        
        return [s for s in splits if s.strip()]
    
    def _create_chunk(
        self,
        text: str,
        document: Document,
        chunk_index: int
    ) -> Chunk:
        """Create a chunk object"""
        chunk_id = f"{document.doc_id}_chunk_{chunk_index}"
        
        return Chunk(
            content=text,
            metadata={
                **document.metadata,
                'chunk_index': chunk_index,
                'chunk_length': len(text)
            },
            chunk_id=chunk_id,
            doc_id=document.doc_id
        )

File: src/main/java/com/ragtutor/chunking/DocumentChunker.java
package com.ragtutor.chunking;

import com.ragtutor.config.ChunkingConfig;
import com.ragtutor.schemas.Chunk;
import com.ragtutor.schemas.Document;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Chunk documents with overlap for better retrieval
 */
@Slf4j
@Component
public class DocumentChunker {
    
    private final int chunkSize;
    private final int chunkOverlap;
    private final String separator;
    
    public DocumentChunker(ChunkingConfig config) {
        this.chunkSize = config.getSize();
        this.chunkOverlap = config.getOverlap();
        this.separator = config.getSeparator();
        log.info("Initialized DocumentChunker with size={}, overlap={}", chunkSize, chunkOverlap);
    }
    
    /**
     * Chunk a document into smaller pieces
     */
    public List<Chunk> chunkDocument(Document document) {
        String text = document.getContent();
        
        // Split by separator first
        List<String> splits = splitText(text, separator);
        
        List<Chunk> chunks = new ArrayList<>();
        List<String> currentChunk = new ArrayList<>();
        int currentLength = 0;
        
        for (String split : splits) {
            int splitLength = split.length();
            
            if (currentLength + splitLength > chunkSize && !currentChunk.isEmpty()) {
                // Create chunk
                String chunkText = String.join(separator, currentChunk);
                Chunk chunk = createChunk(chunkText, document, chunks.size());
                chunks.add(chunk);
                
                // Start new chunk with overlap
                int overlapLength = 0;
                List<String> overlapParts = new ArrayList<>();
                
                for (int i = currentChunk.size() - 1; i >= 0; i--) {
                    String part = currentChunk.get(i);
                    if (overlapLength + part.length() < chunkOverlap) {
                        overlapParts.add(0, part);
                        overlapLength += part.length();
                    } else {
                        break;
                    }
                }
                
                currentChunk = overlapParts;
                currentLength = overlapLength;
            }
            
            currentChunk.add(split);
            currentLength += splitLength;
        }
        
        // Add final chunk
        if (!currentChunk.isEmpty()) {
            String chunkText = String.join(separator, currentChunk);
            Chunk chunk = createChunk(chunkText, document, chunks.size());
            chunks.add(chunk);
        }
        
        log.debug("Chunked document {} into {} chunks", document.getDocId(), chunks.size());
        return chunks;
    }
    
    /**
     * Split text by separator
     */
    private List<String> splitText(String text, String separator) {
        List<String> splits = new ArrayList<>();
        
        if (separator != null && !separator.isEmpty()) {
            String[] parts = text.split(separator);
            for (String part : parts) {
                if (!part.trim().isEmpty()) {
                    splits.add(part);
                }
            }
        } else {
            for (char c : text.toCharArray()) {
                splits.add(String.valueOf(c));
            }
        }
        
        return splits;
    }
    
    /**
     * Create a chunk object
     */
    private Chunk createChunk(String text, Document document, int chunkIndex) {
        String chunkId = document.getDocId() + "_chunk_" + chunkIndex;
        
        Map<String, Object> metadata = new HashMap<>();
        if (document.getMetadata() != null) {
            metadata.putAll(document.getMetadata());
        }
        metadata.put("chunk_index", chunkIndex);
        metadata.put("chunk_length", text.length());
        
        return Chunk.builder()
            .content(text)
            .chunkId(chunkId)
            .documentId(document.getDocId())
            .metadata(metadata)
            .chunkIndex(chunkIndex)
            .build();
    }
}
